cache/contenthash/checksum_test.go

```diff
@@ -434,6 +434,114 @@ func TestChecksumBasicFile(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestChecksumIncludeExclude(t *testing.T) {
+	t.Parallel()
+	tmpdir, err := ioutil.TempDir("", "buildkit-state")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	snapshotter, err := native.NewSnapshotter(filepath.Join(tmpdir, "snapshots"))
+	require.NoError(t, err)
+	cm, _ := setupCacheManager(t, tmpdir, "native", snapshotter)
+	defer cm.Close()
+
+	ch := []string{
+		"ADD foo file data0",
+		"ADD bar file data1",
+		"ADD d0 dir",
+		"ADD d0/abc file abc",
+		"ADD d1 dir",
+		"ADD d1/def file def",
+	}
+
+	ref := createRef(t, cm, ch)
+
+	cc, err := newCacheContext(ref.Metadata(), nil)
+	require.NoError(t, err)
+
+	dgst, err := cc.Checksum(context.TODO(), ref, "foo", ChecksumOpts{IncludePatterns: []string{"foo"}}, nil)
+	require.NoError(t, err)
+
+	require.Equal(t, dgstFileData0, dgst)
+
+	dgstFoo, err := cc.Checksum(context.TODO(), ref, "", ChecksumOpts{IncludePatterns: []string{"foo"}}, nil)
+	require.NoError(t, err)
```

Comment:
I don't think this will match `dgstFileData0` because the record for `/` is included in the checksum (in case other files are added in the future which match the pattern).

Reply:
If source ends with `/` (same for `/`) but then the dir checksum should not apply, correct? As the dir itself is not copied.